File: daily/2025/1353_Maximum_Number_of_Events_That_Can_Be_Attended.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int maxEvents(vector<vector<int>>& events) {
        int ans = 0;
        auto cmp = [](vector<int>& a, vector<int>& b){
            if (a[0] != b[0]) return a[0] < b[0];
            else return a[1] < b[1];
        };
        sort(events.begin(), events.end(), cmp);
        priority_queue<int, vector<int>, greater<int>> pq;
        int n = events.size();
        int idx = 0;
        int t = events[0][0];
        while(idx < n || !pq.empty()){
            while(idx < n && events[idx][0] <= t){
                pq.push(events[idx++][1]);
            }
            while (!pq.empty() && pq.top() < t){
                pq.pop();
            }
            if (!pq.empty() && pq.top() >= t){
                pq.pop();
                ans++;
            }
            t++;
        }
        return ans;
    }
};

int main(){
    vector<vector<int>> events = {{1,2},{2,3},{3,4}};
    // vector<vector<int>> events = {{1,2},{2,3},{3,4},{1,2}};
    Solution S;
    cout << S.maxEvents(events) << endl;
    return 0;
}

File: daily_python/1353_Maximum_Number_of_Events_That_Can_Be_Attended.py
import os
import sys
from typing import List
import heapq

class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        events.sort()
        n = len(events)
        idx = 0
        pq = []
        t = events[0][0]
        ans = 0
        while idx < n or pq:
            while idx < n and events[idx][0] <= t:
                heapq.heappush(pq, events[idx][1])
                idx += 1
            while pq and pq[0] < t:
                heapq.heappop(pq)
            if pq:
                heapq.heappop(pq)
                ans += 1
            t += 1
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            events = [list(map(int, event.split(','))) for event in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append(events)
    return dataset


if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for events in dataset:
        results.append(solution.maxEvents(events))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')